packages/go_router/lib/src/delegate.dart

```diff
@@ -42,22 +43,57 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   /// Set to true to disable creating history entries on the web.
   final bool routerNeglect;
 
-  final GlobalKey<NavigatorState> _key = GlobalKey<NavigatorState>();
-
-  RouteMatchList _matches = RouteMatchList.empty();
+  RouteMatchList _matchList = RouteMatchList.empty();
   final Map<String, int> _pushCounts = <String, int>{};
+  final RouteConfiguration _configuration;
+
+  @override
+  Future<bool> popRoute() async {
+    // Iterate backwards through the RouteMatchList until seeing a GoRoute
+    // with a non-null parentNavigatorKey or a ShellRoute with a non-null parentNavigatorKey
+    // and pop from that Navigator instead of the root.
+    final int matchCount = _matchList.matches.length;
+    for (int i = matchCount - 1; i >= 0; i--) {
+      final RouteMatch match = _matchList.matches[i];
+      final RouteBase route = match.route;
+
+      // If this is a ShellRoute, then pop one of the subsequent GoRoutes, if
+      // there are any.
+      if (route is ShellRoute && (matchCount - i) > 2) {
+        final NavigatorState? navigator = route.navigatorKey.currentState;
+        final bool didPop = await navigator?.maybePop() ?? false;
+        if (didPop) {
+          return didPop;
+        }
+      }
+    }
+
+    // Use the root navigator if no ShellRoute Navigators were found and didn't
+    // pop
+    final NavigatorState? navigator = navigatorKey.currentState;
+
+    if (navigator == null) {
```

Comment:
Can this be null?

Reply:
I'm not sure. It's a nullable type so I don't see a reason to opt-out of the type promotion here.